### PR TITLE
Add option to allow bonus points in assessments

### PR DESCRIPTION
**Pull request by jonatanschroeder**
This option allows an assessment to have a score percentage over 100%.

* [x] Database, schema and sync;
* [x] Sprocs that update assessment score percentage;
* [x] Instructor view updates;
* [x] Student view updates;
* [x] API information about an assessment;
* [x] Docs;
* [x] Example course;
* [x] Tests.

### CODE REVIEW

**Comment by jonatanschroeder** [comment]
Sorry about the moves up there, the GitHub app, when you open the details of a PR, gives you a Close button that I thought would close the details page, but it closes the PR instead without warning.

File: docs/assessment.md [hunk 1]
```diff
@@ -94,11 +94,30 @@ An assessment is broken down in to a list of zones, like this:
...
 
-**`Homework` assessment updates:** New questions added to Homeworks will be automatically integrated into student homeworks currently in progress. Updates to `maxPoints` will take effect the next time a student grades a question. A student's “points” and “percentage score” will never decrease.
+**`Homework` assessment updates:** New questions added to Homeworks will be automatically integrated into student homeworks currently in progress. Updates to `maxPoints` or `maxBonusPoints` will take effect the next time a student grades a question. A student's “points” and “percentage score” will never decrease.
+
+## Assessment points
+
+A student's percentage score will be determined by the number of points they have obtained, divided by the `maxPoints` setting.
```
Comment (nwalters512) [inline docs/assessment.md:105]: Nit: I'd prefer to use "value" or "option" in place of "setting" to refer to these.

File: docs/assessment.md [hunk 2]
```diff
@@ -94,11 +94,30 @@ An assessment is broken down in to a list of zones, like this:
...
 
-**`Homework` assessment updates:** New questions added to Homeworks will be automatically integrated into student homeworks currently in progress. Updates to `maxPoints` will take effect the next time a student grades a question. A student's “points” and “percentage score” will never decrease.
+**`Homework` assessment updates:** New questions added to Homeworks will be automatically integrated into student homeworks currently in progress. Updates to `maxPoints` or `maxBonusPoints` will take effect the next time a student grades a question. A student's “points” and “percentage score” will never decrease.
```
Comment (nwalters512) [inline docs/assessment.md:101]: Nit: consistent quotes with the rest of the document
```suggestion
**`Homework` assessment updates:** New questions added to Homeworks will be automatically integrated into student homeworks currently in progress. Updates to `maxPoints` or `maxBonusPoints` will take effect the next time a student grades a question. A student's "points" and "percentage score" will never decrease.
```

File: docs/assessment.md (hunk 1 above)
Comment (nwalters512) [inline docs/assessment.md:105]: Long term we should ideally unify the documentation around points, but in the meantime I'd recommend cross-linking with the Credit section on the Access Control page.

EDIT: I see the link to that page at the end of this section - would consider lifting it up to the top of the section to make it more visible.

File: docs/assessment.md [hunk 3]
```diff
@@ -94,11 +94,30 @@ An assessment is broken down in to a list of zones, like this:
...
+    "uuid": "cef0cbf3-6458-4f13-a418-ee4d7e7505dd",
+    "maxPoints": 50,
+    "maxBonusPoints": 5,
+    ...
+}
+```
+
+The `maxPoints` determines the number of points a student is required to obtain to get a score of 100%. The percentage score will thus be computed based on the points the student obtained divided by this setting. If not provided, `maxPoints` is computed based on the maximum number of points that can be obtained from all questions in all zones.
+
+By default, once a student obtains enough points to reach the value of `maxPoints`, any further points do not affect the assessment score. If, however, the `maxBonusPoints` setting is used, the student can obtain additional points, up to a total of `maxPoints + maxBonusPoints`. The percentage is still based on `maxPoints`, so the use of `maxBonusPoints` allows students to obtain a percentage above 100%. If `maxBonusPoints` is set, but `maxPoints` is not provided, then `maxPoints` will be computed by subtracting `maxBonusPoints` from the maximum number of points in all questions.
```
Comment (nwalters512) [inline docs/assessment.md:118]:
```suggestion
By default, once a student obtains enough points to reach the value of `maxPoints`, any further points do not affect the assessment score. However, if the `maxBonusPoints` setting is used, the student can obtain additional points, up to a total of `maxPoints + maxBonusPoints`. The percentage is still based on `maxPoints`, so the use of `maxBonusPoints` allows students to obtain a percentage above 100%. If `maxBonusPoints` is set, but `maxPoints` is not provided, then `maxPoints` will be computed by subtracting `maxBonusPoints` from the maximum number of points in all questions.
```

File: sprocs/assessment_instances_update.sql [hunk 4]
```diff
@@ -102,24 +107,24 @@ BEGIN
...
-        SELECT
-            sum(zmp.max_points)
-        INTO
-            new_assessment_instance_max_points
-        FROM
-            assessment_instances_points(assessment_instance_id) AS zmp;
-    END IF;
+    new_assessment_instance_max_points := COALESCE(assessment_max_points, GREATEST(zone_max_points - assessment_max_bonus_points, 0));
+
+    -- ensure that max_bonus_points is at most based on available points
```
Comment (nwalters512) [inline sprocs/assessment_instances_update.sql:117]: Nit: not sure I fully understand this comment - can you reword?

File: sprocs/assessment_instances_update.sql [hunk 5]
```diff
@@ -102,24 +107,24 @@ BEGIN
     -- did we add any instance questions above?
     updated := updated OR (cardinality(new_instance_question_ids) > 0);
 
+    SELECT sum(zmp.max_points)
+    INTO zone_max_points
```
Comment (nwalters512) [inline sprocs/assessment_instances_update.sql:111]: Can we name this `zones_total_max_points` or the like (bikeshed away) to indicate that this refers to a sum across all zones and not a single zone? This'll make the later calculation (`GREATEST(zone_max_points - assessment_max_bonus_points, 0)`) a bit clearer.

File: tests/testBonusPoints.js [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+            assessment_id: context.assessmentId,
+        };
+        const results = await sqldb.queryOneRowAsync(sql.read_assessment_instance_points, params);
+        assert.equal(results.rowCount, 1);
+        // 6+8 is 14, but limit should be 10+2 (max plus bonus)
+        assert.equal(results.rows[0].points, 12);
+        assert.equal(results.rows[0].score_perc, 120);
+    });
+    
+
```
Comment (nwalters512) [inline tests/testBonusPoints.js:106]: formatting nit:
```suggestion
```

File: sprocs/assessment_instances_grade.sql [hunk 7]
```diff
@@ -75,27 +76,24 @@ BEGIN
...
-    SELECT ai.score_perc INTO current_score_perc
+    SELECT ai.max_points, ai.max_bonus_points, ai.score_perc
+    INTO max_points, max_bonus_points, current_score_perc
     FROM assessment_instances AS ai
     WHERE ai.id = assessment_instance_id;
 
     -- #########################################################################
     -- awarded points and score_perc
 
     -- compute the score in points, maxing out at max_points
```
Comment (nwalters512) [inline sprocs/assessment_instances_grade.sql:87]:
```suggestion
    -- compute the score in points, maxing out at max_points + max_bonus_points
```

**Review by nwalters512 — APPROVED** [review]
Sorry for all the nits - grading is some of the most complex and potentially confusing stuff in PrairieLearn, so I want to make sure it's as clear and documented as possible! This looks good on the whole though.

I won't merge this quite yet in case Matt gets a chance to look later today - I'd trust his review of grading code more than my own 🙂

**Pull request merged**